### ISSUE TITLE
[$250] Web - Workspace - invite button at the invite new members page does not prevent a user from multiple button press @jayeshmangwani

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#9270]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Navigate to workspace
2. Go to manage members > Invite
3. Select user/s and press the invite button twice
4. It's adding a user 2 times and shows a key warning

## Expected Result:

There should not be an option to add a given user twice; page should remain on the manage member page (rather than moving to workspace screen page)

## Actual Result: 

Invite button add user to workspace twice, navigates to the workspace page

## Workaround:

Unknown

## Platform:

Where is this issue occurring?

- Web

**Version Number:** 1.1.70-1

**Reproducible in staging?:** Y

**Reproducible in production?:** Y

**Email or phone of affected tester (no customers):** 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/31225194/172389196-aabcf5a3-572a-4312-8001-cd0cdc07cf1e.mov

https://user-images.githubusercontent.com/43996225/171425785-93b92ca5-c424-4ef6-bcc0-0dced8aa8b98.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @jayeshmangwani
**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by stephanieelliott** [comment]
Hey @kavimuru I'm not sure if I understand the issue here. 

Can you clarify on: 

1. What do you mean when you say it's adding a user 2 times? 

   - When I click on Invite > [select invitee] > Invite, it takes me back to the `Manage Members` screen (same as in the video), and the invitee appears only once on the list. If they were added 2 times, I would expect them to appear in the list twice - am I looking in the wrong place? 

2. What does `shows a key warning` mean in this context? 
   - I'm not seeing any errors, warnings or modals when I take the action steps. Can you point me to the timestamp in the video where you see a ket warning?

**Comment by kavimuru** [comment]
@stephanieelliott  1. The issue is reproducible inconsistently. Sometimes user can click on "invite" button more than once and if we click couple times the invitee is added twice. 
2. Key warning is not reproducible at all. Please ignore that step.

**Comment by jayeshmangwani** [comment]
https://user-images.githubusercontent.com/35371050/172000967-719bf382-782b-45d5-91fb-687e38057fea.mov

@stephanieelliott @kavimuru 
1. in the invite new members page select a user and press the invite button 2 times, it will add a user twice, either we have to prevent multiple touches on the page or need to handle the same user can not be added twice
2. please ignore key warnings, it's getting when we are running debug app (as we can see this warning in this video)

**Comment by stephanieelliott** [comment]
Thanks @jayeshmangwani, that makes sense now! Updated the issue description and added your video to the issue body. I think there is value in fixing this so will move it through triage.

**Comment by Gonals** [comment]
Setting as external!

**Comment by anjumdar7864** [comment]
Hi, as i understand the issue. when we do listing every row has unique key. so as i understand if key is coming from api then we have to get distinct keys. other than that if key is being set from loop variable then its loop variable issue. i am interested to fix this bug. i am FULL STACK developer and i have more than 5 years experience.

**Comment by JosueEchandiaAsto** [comment]
### Proposal

### Issue

Invite button add user to workspace twice, navigates to the workspace page. When press the invite button twice.

### Fix

Each time you press the button, the invitation service is consulted. 
So what we should do is disable the button after pressing the button. 

For this we will add `isLoadingInviteToWorkspace` property, it will be a boolean that will return true if it is consulting the invitation service.
Create `IS_LOADING_INVITE_TO_WORKSPACE` in `ONYXKEYS`
https://github.com/Expensify/App/blob/f55c2c5ad98f15a8f3b5f9e93af45d201ee55b37/src/ONYXKEYS.js#L176-L177
Just before calling the service `IS_LOADING_INVITE_TO_WORKSPACE` will be true
https://github.com/Expensify/App/blob/f55c2c5ad98f15a8f3b5f9e93af45d201ee55b37/src/libs/actions/Policy.js#L339-L348
When the service call completes `IS_LOADING_INVITE_TO_WORKSPACE` will be false
https://github.com/Expensify/App/blob/f55c2c5ad98f15a8f3b5f9e93af45d201ee55b37/src/libs/actions/Policy.js#L373
Commit: https://github.com/Expensify/App/commit/f55c2c5ad98f15a8f3b5f9e93af45d201ee55b37

In `WorkspaceInvitePage.js`
Add property `isLoadingInviteToWorkspace` https://github.com/Expensify/App/blob/f55c2c5ad98f15a8f3b5f9e93af45d201ee55b37/src/pages/workspace/WorkspaceInvitePage.js#L54-L55
Set default value of `isLoadingInviteToWorkspace` 
https://github.com/Expensify/App/blob/f55c2c5ad98f15a8f3b5f9e93af45d201ee55b37/src/pages/workspace/WorkspaceInvitePage.js#L61-L64
Get property with Onyx
https://github.com/Expensify/App/blob/f55c2c5ad98f15a8f3b5f9e93af45d201ee55b37/src/pages/workspace/WorkspaceInvitePage.js#L364-L373

Use of `isLoadingInviteToWorkspace`:

1. When the service is loading the FormAlertWithSubmitButton is disabled. https://github.com/Expensify/App/blob/f55c2c5ad98f15a8f3b5f9e93af45d201ee55b37/src/pages/workspace/WorkspaceInvitePage.js#L322-L324
2. When the service is loading you cannot modify the list of selected users to invite
https://github.com/Expensify/App/blob/f55c2c5ad98f15a8f3b5f9e93af45d201ee55b37/src/pages/workspace/WorkspaceInvitePage.js#L273-L280

### Result

https://user-images.githubusercontent.com/16966456/174421068-d45c1ba8-45dc-4cd2-ac6c-0bf5acef8ec2.mp4


https://user-images.githubusercontent.com/16966456/174421146-91b7ce6e-639f-4296-ab66-d3a9d2713937.mov

**Comment by YoucefBen47** [comment]
The fix for this issue should be simple. We create a piece of state and set it to false. We use this state to determin whether the button is disabled or not. Once the button is clicked we change the state from false to true and make the button disabled (unclickable).

**Comment by Julesssss** [comment]
Hi @JosueEchandiaAsto, thanks for the proposal. 

I'm going to put this on hold for a bit as we're in the middle of refactoring our API (which changes your solution). Let's return to this in a couple of weeks.

**Comment by Julesssss** [comment]
Our refactor is still in progress. More info soon!

**Comment by Julesssss** [comment]
Still no progress

**Comment by Julesssss** [comment]
Update to follow soon

**Comment by mallenexpensify** [comment]
Still on hold

**Comment by mallenexpensify** [comment]
@puneetlath Reassigning as I'm heading OOO for a month

**Comment by puneetlath** [comment]
@Julesssss should this still be on hold?

**Comment by Julesssss** [comment]
Here was [the context](https://github.com/Expensify/App/issues/9270#issuecomment-1160462001) (API Refactor). Seems like this can be un-held now though.

**Comment by jayeshmangwani** [comment]
@puneetlath Yes it looks like the issue is resolved now, I am also not able to reproduce this issue

**Comment by puneetlath** [comment]
Ok, looks like the API refactor project fixed it. @jayeshmangwani if you apply here, we can get you paid out for the reporting bonus and then close this out. https://www.upwork.com/jobs/~015b629c650416fcd5

**Comment by jayeshmangwani** [comment]
@puneetlath Applied, Thanks
